### ISSUE TITLE
select with IPv4 type results in "Type mismatch in IN or VALUES section: IPv4. Got: UInt64:" error

### ISSUE THREAD

**Issue by bharatnc** [ClickHouse/ClickHouse#54160]
**Describe the unexpected behaviour**
For example selecting with toIPv4 results in the following error:
```
:) select toIPv4('44.44.44.44') IN IPv4StringToNum('1.1.1.1');

SELECT toIPv4('44.44.44.44') IN IPv4StringToNum('1.1.1.1')

0 rows in set. Elapsed: 0.001 sec.

Received exception from server (version 23.3.9):
Code: 53. DB::Exception: Received from <redacted>:9440. DB::Exception: Type mismatch in IN or VALUES section. Expected: IPv4. Got: UInt64: While processing toIPv4('44.44.44.44') IN IPv4StringToNum('1.1.1.1'). (TYPE_MISMATCH)
```
Current workaround is to convert it to UInt64 using `toUInt64`:
```
select toUInt64(toIPv4('44.44.44.44')) IN IPv4StringToNum('1.1.1.1');
```

**How to reproduce**
To reproduce you can use this minimally reproducible query:
```
select toIPv4('44.236.15.91') IN IPv4StringToNum('1.1.1.1');
```
* Which ClickHouse server version to use: `v23.3-lts` as well as `latest upstream source build`

**Expected behavior**

I would expect this query to pass with v23.3-lts. Works correctly with v22.3-lts.
```
# With v22.3-lts it works fine

select toIPv4('44.44.44.44') IN IPv4StringToNum('1.1.1.1');

SELECT toIPv4('44.44.44.44') IN IPv4StringToNum('1.1.1.1')

┌─in(toIPv4('44.44.44.44'), IPv4StringToNum('1.1.1.1'))─┐
│                                                      0 │
└────────────────────────────────────────────────────────┘

1 rows in set. Elapsed: 0.001 sec.

```
**Error message and/or stacktrace**
```
# with v23.3-lts
:) select toIPv4('44.44.44.44') IN IPv4StringToNum('1.1.1.1');

SELECT toIPv4('44.44.44.44') IN IPv4StringToNum('1.1.1.1')

0 rows in set. Elapsed: 0.001 sec.

Received exception from server (version 23.3.9):
Code: 53. DB::Exception: Received from <redacted>:9440. DB::Exception: Type mismatch in IN or VALUES section. Expected: IPv4. Got: UInt64: While processing toIPv4('44.44.44.44') IN IPv4StringToNum('1.1.1.1'). (TYPE_MISMATCH)
```
**Additional context**

This is a minimal reproducible example. For more context, upgrading from v22.3-lts to v23.3-lts results in such an error when selecting from a table column that has IPv4 type and using IN ('1.1.1.1') clause.

For example:
```
-- addressIPv4 is a column with IPv4 type
select addressIPv4 IN ('1.1.1.1') from foo limit 1;
```
Seems loosely related to: `https://github.com/ClickHouse/ClickHouse/pull/43221` and looks like there is no handling for `if (which_type.isIPv4() && src.getType() == Field::Types::UInt64)` which seems to fall through and cause the error here: https://github.com/ClickHouse/ClickHouse/blob/master/src/Interpreters/convertFieldToType.cpp#L518. Not sure what's the right fix here. 

Thank you in advance for looking into this issue.

**Comment by bharatnc** [comment]
cc: @yakov-olkhovskiy

**Comment by yakov-olkhovskiy** [comment]
We moved from domain IP types to native in v23, so this behavior is expected.
`select addressIPv4 IN ('1.1.1.1') from foo` should work:
https://fiddle.clickhouse.com/d8fdaf44-a77f-4d2a-925b-761b986af536

**Comment by den-crane** [comment]
IPv4StringToNum produces type Int32
toIPv4 produces type IPv4.

It worked accidentally, that IPv4 stored data similar way and had implicit cast. 
But these two types could store bytes in different order.

**Comment by bharatnc** [comment]
yeah, here is a way to reproduce it: https://fiddle.clickhouse.com/3e663833-263d-4008-bad2-d852872c6a20

**Comment by bharatnc** [comment]
fixed in #54212
